### PR TITLE
Enhancement of class MutableAsciiBuffer method putAsciiFloat - leading zeros

### PR DESCRIPTION
**Pull request by ariksher-nex**
solves issue #227

### CODE REVIEW

**Comment by RichardWarburton** [comment]
Thanks for the PR, but it has caused the build to fail - please run `./gradlew` and fix the checkstyle errors.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java [hunk 1]
```diff
@@ -45,6 +47,30 @@
...
+        10_000_000_000L,
+        100_000_000_000L,
+        1_000_000_000_000L, // scale=12
+        10_000_000_000_000L,
+        100_000_000_000_000L,
+        1_000_000_000_000_000L, // scale=15
+        10_000_000_000_000_000L,
+        100_000_000_000_000_000L,
+        1_000_000_000_000_000_000L, // scale=18
+        //10_000_000_000_000_000_000L, // scale=19
```
Comment (RichardWarburton) [inline artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: Why is this commented out?

File: artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java [hunk 2]
```diff
@@ -48,6 +48,8 @@ public void canEncodeDecimalFloat()
     {
         final DecimalFloat price = new DecimalFloat(value, scale);
 
-        assertEquals(input, price.toString());
+        // since toString of DecimalFloat does not return leading zeros, it fails
+        //assertEquals(input, price.toString());
```
Comment (RichardWarburton) [inline artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java]: Why does this PR just comment out asserts? The the output of the `toString()` method should be consistent with printing the number on the `AsciiBuffer`.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java [hunk 3]
```diff
@@ -36,27 +36,92 @@
...
             {"1.00109125", 100109125, 8},
             {"-1.00109125", -100109125, 8},
+
+            // negative scale
+            {"2000", 2L, -3},
+            {"-2000", -2L, -3},
+
+            // edge values
+            {"0", 0, 0},
+            //{"0.00", 0, 2},
```
Comment (RichardWarburton) [inline artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java]: Why is this value commented out? Either it should pass or if it isn't a valid value then it should be removed.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java (hunk 1 above)
Comment (ariksher-nex) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: cannot really represent the value of 10 * 10^18.
look at method **checkForLeadingZeroAndPutIfNeeded** and see that if scale is GE 19 there is special treatment.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java [hunk 4]
```diff
@@ -617,6 +644,16 @@ private long calculateRemainderAndPutMinus(final int offset, final long value)
         }
     }
 
+    private int checkForLeadingZeroAndPutIfNeeded(final int offset, final long value, final int scale)
```
Comment (RichardWarburton) [inline artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: Is this the most efficient way to do this? Can't we just check if the number of zeros is > 0?

**Comment by RichardWarburton** [comment]
I've added a few inline comments.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java (hunk 2 above)
Comment (ariksher-nex) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java]: I left a comment so you could see it.  :-)
**must** it be consistent?
look at line below I use Float.valueOf.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java (hunk 1 above)
Comment (RichardWarburton) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: In that case it's probably better to just remove the scale=19 entry.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java (hunk 3 above)
Comment (RichardWarburton) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java]: I would just remove this entry. A value of 0 is always equal to zero and there's point in printing out "0.00" when "0" would suffice.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java (hunk 2 above)
Comment (RichardWarburton) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java]: I think it should be consistent, you don't want to see a value in your debugger that is different from the one printed out on the buffer. The implementation of toString() does not have to be fast though as people should never use that on the critical path of their application.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java (hunk 4 above)
Comment (ariksher-nex) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: I tried to minimize the code change.
so i treated the leading zero similar to the minus sign.
if going to refer the zeros count it would lead to larger code change.
why do you think it is not efficient? the access to the array is O(1).

Comment (RichardWarburton) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: Well, given that you had the number of zeros in a local field, It just looked to me you could do a quick check on that. Perhaps I'm miss-reading the diff ;) I don't think your approach is bad, just wondering whether it wasn't going the other route?

Comment (ariksher-nex) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: well I managed to figure it out your way. take a look at new code.

Comment (RichardWarburton) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: Great, thanks.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java (hunk 2 above)
Comment (ariksher-nex) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java]: i tried to make it consistent, than encountered other test that fails...
test class **DecimalFloatToAndFromDoubleTest** method **shouldConvertFromDoubleAccurately**

> uk.co.real_logic.artio.fields.DecimalFloatToAndFromDoubleTest > shouldConvertFromDoubleAccurately FAILED
>     org.junit.ComparisonFailure: expected:<[].01> but was:<[0].01>
>         at org.junit.Assert.assertEquals(Assert.java:115)
>         at org.junit.Assert.assertEquals(Assert.java:144)
>         at uk.co.real_logic.artio.fields.DecimalFloatToAndFromDoubleTest.shouldConvertFromDoubleAccurately(DecimalFloatToAndFromDoubleTest.java:45)

as I understand the code, the consistency between double/float of java lang and DecimalFloat is **more important** than consistency of DecimalFloat to MutableAsciiBuffer.
if that is correct the test of DecimalFloatToStringTest.java might be obsolete.

Comment (ariksher-nex) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java]: ok made it consistent. they both output with leading zeros.
had to modify test class **DecimalFloatToAndFromDoubleTest** accordingly.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java (hunk 1 above)
Comment (ariksher-nex) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: removed entirely this block.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java (hunk 3 above)
Comment (ariksher-nex) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java]: ok removed it.

File: artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java (hunk 1 above)
Comment (ariksher-nex) [reply artio-codecs/src/main/java/uk/co/real_logic/artio/util/MutableAsciiBuffer.java]: I removed the entire block, with simpler implementation.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java (hunk 3 above)
Comment (ariksher-nex) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/util/DecimalFloatEncodingTest.java]: ok, removed it.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java (hunk 2 above)
Comment (RichardWarburton) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/fields/DecimalFloatToStringTest.java]: Thanks.

**Comment by RichardWarburton** [comment]
@ariksher-nex - change looks good to me now, but the build is failing on CI - looks like checkstyle.

**Comment by ariksher-nex** [comment]
@RichardWarburton fixed the checkstyle.
